### ISSUE TITLE
How to reproduce the result of abstractive document summarization on Transformer?

### ISSUE THREAD

**Issue by yuta-hitomi** [pytorch/fairseq#1012]
Hi

I read https://arxiv.org/pdf/1904.01038.pdf. I want to reproduce the results in the third row of Table 5. How to set hyper-parameters?

**Comment by Michaelvll** [comment]
Hi

I am having the same problems, too. Did you find out what the configuration is? :)

**Comment by yuta-hitomi** [comment]
I used this script to reproduce the result.
```
python train.py $TEXT/data.src_trg \
    --ddp-backend=no_c10d \
    --arch transformer_wmt_en_de --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
    --lr-scheduler inverse_sqrt --warmup-init-lr 1e-07 --warmup-updates 4000 \
    --lr 0.0007 --min-lr 1e-09 \
    --criterion label_smoothed_cross_entropy --label-smoothing 0.1 --weight-decay 0.0 \
    --max-tokens  4096 --dropout 0.1 \
    --skip-invalid-size-inputs-valid-test \
    --update-freq 8 --no-progress-bar --log-format json --log-interval 50 \
    --save-interval-updates  1000 --keep-interval-updates 20 \
    --no-epoch-checkpoints \
    --max-epoch 30 \
    --save-dir $TEXT/models
```
I followed the instruction.
https://github.com/pytorch/fairseq/issues/346#issuecomment-435652674

Our results:
```
ROUGE-1: 40.8
ROUGE-2: 18.3
ROUGE-L: 37.7
```

**Comment by yuta-hitomi** [comment]
Sorry, we use the first 1000 tokens as input.

**Comment by Michaelvll** [comment]
Thank you so much for your reply. 
Could you please be so kind as to share the script for preprocessing the CNN Dailymail dataset? I keep getting very low ROUGE scores (around 0.001).

B.t.w, I used the following code for preprocessing and testing.
Preprocessing:
```bash
src=src
tgt=tgt

BPEROOT=subword-nmt
BPE_TOKENS=30000

prep=data/cnn/cnndm_30k
tmp=$prep/tmp

TRAIN=$tmp/train.txt.src-tgt
BPE_CODE=$prep/code
rm -f $TRAIN
for l in $src $tgt; do
    cat $tmp/train.txt.$l >> $TRAIN
done

echo "learn_bpe.py on ${TRAIN}..."
python $BPEROOT/learn_bpe.py -s $BPE_TOKENS < $TRAIN > $BPE_CODE

for L in $src $tgt; do
    for f in train.txt.$L val.txt.$L test.txt.$L; do
        echo "apply_bpe.py to ${f}..."
        python $BPEROOT/apply_bpe.py -c $BPE_CODE < $tmp/$f > $prep/$f
    done
done

TEXT=$prep
python configs/cnn_daily/truncate.py  --source-lang src --target-lang tgt \
    --trainpref $TEXT/train.txt --validpref $TEXT/val.txt --testpref $TEXT/test.txt --trunc-size 400

fairseq-preprocess --source-lang src --target-lang tgt \
    --trainpref $TEXT/train.txt.trunc --validpref $TEXT/val.txt.trunc --testpref $TEXT/test.txt.trunc \
    --destdir data/binary/cnndm_30k \
    --workers 20 --joined-dictionary
```
Testing:
```bash
checkpoint=$1
output_path=$(dirname -- "$checkpoint")/exp
mkdir -p $output_path

python generate.py data/binary/cnndm_30k --path $checkpoint --remove-bpe --gen-subset test \
   --batch-size 30 --min-len 35 --max-len-b 200  --beam 5 --lenpen 0.9 --no-repeat-ngram-size 3 > $output_path/cnn_dailymail.out

GEN=$output_path/cnn_dailymail.out
SYS=$GEN.sys
REF=$GEN.ref
grep ^H $GEN | cut -f3- > $SYS
grep ^T $GEN | cut -f2- > $REF

# This requires files2rouge to be installed: https://github.com/pltrdy/files2rouge
files2rouge $SYS $REF

```

**Comment by yuta-hitomi** [comment]
This is my script for preprocess.
In evaluation, we use official ROUGE script.

```
find $RAW -name 'train*' | xargs cat | subword-nmt learn-bpe -s 30000 -o $BPE/codecs
for FILE in `find $RAW/*`; do
       NAME=`echo $FILE | sed -e "s*$RAW**"`
       subword-nmt apply-bpe -c $BPE/codecs < $FILE > $BPE/$NAME
done
python $FAIR/preprocess.py \
       --joined-dictionary \
       --source-lang src \
       --target-lang tgt \
       --trainpref $BPE/train \
       --validpref $BPE/valid \
       --destdir $BPE/data.src_tgt \
       --workers 30
```

I think that there are some bugs in your code before the preprocess.

**Comment by apeterswu** [comment]
> This is my script for preprocess.
> In evaluation, we use official ROUGE script.
> 
> ```
> find $RAW -name 'train*' | xargs cat | subword-nmt learn-bpe -s 30000 -o $BPE/codecs
> for FILE in `find $RAW/*`; do
>        NAME=`echo $FILE | sed -e "s*$RAW**"`
>        subword-nmt apply-bpe -c $BPE/codecs < $FILE > $BPE/$NAME
> done
> python $FAIR/preprocess.py \
>        --joined-dictionary \
>        --source-lang src \
>        --target-lang tgt \
>        --trainpref $BPE/train \
>        --validpref $BPE/valid \
>        --destdir $BPE/data.src_tgt \
>        --workers 30
> ```
> 
> I think that there are some bugs in your code before the preprocess.

Hi, could you please kindly provide the script for generation? What's the beam, min-len or related hyperparameters? 
Thanks very much.

**Comment by Michaelvll** [comment]
> This is my script for preprocess.
> In evaluation, we use official ROUGE script.
> 
> ```
> find $RAW -name 'train*' | xargs cat | subword-nmt learn-bpe -s 30000 -o $BPE/codecs
> for FILE in `find $RAW/*`; do
>        NAME=`echo $FILE | sed -e "s*$RAW**"`
>        subword-nmt apply-bpe -c $BPE/codecs < $FILE > $BPE/$NAME
> done
> python $FAIR/preprocess.py \
>        --joined-dictionary \
>        --source-lang src \
>        --target-lang tgt \
>        --trainpref $BPE/train \
>        --validpref $BPE/valid \
>        --destdir $BPE/data.src_tgt \
>        --workers 30
> ```
> 
> I think that there are some bugs in your code before the preprocess.

Hi, @yuta-hitomi, thank you so much for providing the preprocessing code and the training code. I retrained the model with the training script on the data from tensor2tensor repository. I got similar results on average_F of Rouge-1 and Rouge-2, but the Rouge-L is weirdly low (only 29.2). I am wondering if you have any idea about why this happens?
Thank you!

**Comment by danielvarab** [comment]
Here I am more than a year later. I am also struggling to replicate these results. Did anyone figure out if there were any discrepancies with the generation configuration? For references, perplexity and loss seem reasonable, however, ROUGE scores are about 5 points lower across the board.
